### DISCUSSION TITLE
How do I ignore entire file when using v8 converage?

### DISCUSSION THREAD

**Discussion by Kolobamanacas** [vitest-dev/vitest#3739] [Q&A] [answered]
The official docs advise to use `@vitest/coverage-v8` for testing coverage as `c8` will be deprecated in the future. But I can't find anything about excluding files from coverage test when using `@vitest/coverage-v8`.

It is probably important to mention that I don't wish to change Vitest config every time new files are added because in particular cases I still wish to test certain files with the same naming rules. So masking excluded files with regexp/glob in config is not an option. Rather I wish to have an option to manually mark files as to be ignored from coverage, something like Istanbul's `/* istanbul ignore file */`. Is it possible with `v8`?

**Comment by AriPerkkio** [comment]
There is a section for this here https://vitest.dev/guide/coverage.html#ignoring-code. `v8` related ignore hints are 100% identical with `c8`: https://github.com/istanbuljs/v8-to-istanbul#ignoring-uncovered-lines.

The underlying dependency `v8-to-istanbul` does not have `ignore file` hint. Does it work if you add just a `/* c8 ignore start */` on top of the file?

**Reply by Kolobamanacas** [reply]
Something interesting happens when I use `c8 ignore start`. I've created a simple file not covered with tests:

```typescript
export const doMath = (numberA: number, numberB: number): number => {
  return numberA + numberB;
}
```

When I run coverage test as is, it fairly shows 0% coverage in all 4 fields (Stmts, Branch, Funcs and Lines).

But when I run it with `/* c8 ignore start */` in the first line, despite there is no single test for a file, Branch and Funcs show 100% coverage, yet Statements and Lines don't.

![CoverageWithIgnoreStart](https://github.com/vitest-dev/vitest/assets/20444060/a009ab20-6b36-46db-a062-8095e785bd56)

So it works in a way, but not the way I expect and the goal of ignoring the whole file is not achieved.

**Reply by Kolobamanacas** [reply]
I also tested `/* c8 ignore start */` ... `/* c8 ignore stop */` pair embracing the whole file content with the same outcome.

**Reply by AriPerkkio** [reply]
Looks like a bug in `v8-to-istanbul`. The statements are not marked as covered when ignore hint is used.

<details>

```js
import v8ToIstanbul from "v8-to-istanbul";
import { writeFileSync } from "node:fs";

const source = `
/* c8 ignore next 3 */
export const doMath = (numberA: number, numberB: number): number => {
  return numberA + numberB;
};
`.trim();

writeFileSync("test-file.js", source, "utf-8");

const converter = v8ToIstanbul("./test-file.js", 0, { source });

await converter.load();

converter.applyCoverage([
  {
    functionName: "(empty-report)",
    ranges: [
      {
        startOffset: 0,
        endOffset: source.length,
        count: 0,
      },
    ],
    isBlockCoverage: true,
  },
]);

const coverageMap = converter.toIstanbul();
const [fileCoverage] = Object.values(coverageMap);
const { b, f, s } = fileCoverage;

console.log({ b, f, s });

/*
$ node coverage-test.mjs
{
  b: { '0': [ 1 ] },
  f: { '0': 1 },
  s: { '0': 0, '1': 0, '2': 0, '3': 0 }
}
*/
```

</details>

I have a possible fix for this. Unfortunately there are no work-arounds that could be used in Vitest. Only way is to use `coverage.exclude` for now.

**Reply by Kolobamanacas** [reply]
I see, thank you. I will temporary stick to `coverage.exclude` then. But meanwhile I would like to track the progress. Should I report an issue in `v8-to-istanbul` repo then? Or is there any other way I can be of help to speed up the process?

**Reply by AriPerkkio** [reply, marked as answer]
This can be tracked here: https://github.com/istanbuljs/v8-to-istanbul/issues/217

**Reply by AriPerkkio** [reply]
This is now fixed and released. Update `v8-to-istanbul` from your package manager's lockfiles to get latest version.

**Reply by Kolobamanacas** [reply]
Thanks Ari, you are the best! :)
